Consolidate all platforms that have tests

Swamp-Ig opened the pull request:
## Description:

Move all platforms that *have tests* to their own folders.

**Related PR:** completes #22096

Modified script from above PR:
```python
import pathlib
import subprocess
import re
import pytest

path = pathlib.Path('./homeassistant/components/')
test_path = pathlib.Path('./tests/components/')

# Move platforms to proper component dir
whitelist = [
    'remote',
    'air_quality',
    'alarm_control_panel',
    'binary_sensor',
    'calendar',
    'camera',
    'climate',
    'cover',
    'device_tracker',
    'fan',
    'geo_location',
    'image_processing',
    'light',
    'lock',
    'mailbox',
    'media_player',
    'scene',
    'sensor',
    'switch',
    'vacuum',
    'water_heater',
    'weather',
]

# To limit the script to specific integrations
focus_integration = {}


for component in whitelist:
    tests = []

    for fil in (path / component).glob('*.py'):
        if (fil.name in (
                '__init__.py', 'const.py', 'util.py', 'group.py',
                'homeassistant.py', 'reproduce_state.py', 'demo.py')
                or re.match(r'generic_\w+.py', fil.name)):
            continue

        platform = fil.name.rsplit('.', 1)[0]

        if focus_integration and platform not in focus_integration:
            continue

        comp_path = path / platform
        comp_test_path = test_path / platform
        
        # Skip all platforms that don't have tests
        comp_test = test_path / component / f"test_{platform}.py"
        if not comp_test.is_file():
            continue

        if not comp_path.is_dir():
            comp_path.mkdir()
            init = comp_path / '__init__.py'
            with open(init, 'w') as fp:
                fp.write(
                    f"""\"\"\"The {platform} component.
"""
                )
            subprocess.run(f'git add {init}', shell=True)
            
        comp = comp_path / f"{component}.py"
        subprocess.run(f'git mv {fil} {comp}',
                       shell=True)

        if not comp_test_path.is_dir():
            comp_test_path.mkdir()
            init = comp_test_path / '__init__.py'
            with open(init, 'w') as fp:
                fp.write(
                    f"""\"\"\"Tests for the {platform} component.\"\"\"
"""
                )
            subprocess.run(f'git add {init}', shell=True)

        test_target = comp_test_path / f"test_{component}.py"
        subprocess.run(f'git mv {comp_test} {test_target}', shell=True)

        tests.append((test_target, platform))

    subprocess.run(f'git commit -m "Moved {component} components with tests into platform dirs."', shell=True)

    fail = False

    for (test_target, platform) in tests:
        with open(test_target) as file:
            contents = file.read()

        contents = re.sub(
            f'from homeassistant.components.{component} import {platform}',
            f'from homeassistant.components.{platform} import {component} as {platform}',
            contents)
        contents = re.sub(
            f'from homeassistant.components.{component}.{platform}',
            f'from homeassistant.components.{platform}.{component}',
            contents)
        contents = re.sub(
            f'import homeassistant.components.{component}.{platform} as (.*)',
            rf'import homeassistant.components.{platform}.{component} as \1',
            contents)
        contents = re.sub(
            f'import homeassistant.components.{component}.{platform}',
            f'import homeassistant.components.{platform}.{component} as {platform}',
            contents)
        contents = re.sub(
            rf'patch\(\'homeassistant.components.{component}.{platform}',
            f"patch('homeassistant.components.{platform}.{component}",
            contents)

        with open(test_target, 'w') as file:
            file.write(contents)

        if pytest.main([str(test_target)]) == 0:
            subprocess.run(f'git add {test_target}', shell=True)
        else:
            fail = True

    if fail:
        print(f'Warning: failed test rewrite in component {component}. Fix this, add the target and commit. Rerun the script.')
        exit(1)

    subprocess.run(f'git commit -m "Updated tests from {component} component."', shell=True)
```

Swamp-Ig committed:
Moved climate components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from climate component.

Swamp-Ig committed:
Moved binary_sensor components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from binary_sensor component.

Swamp-Ig committed:
Moved calendar components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from calendar component.

Swamp-Ig committed:
Moved camera components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from camera component.

Swamp-Ig committed:
Moved cover components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from cover component.

Swamp-Ig committed:
Moved device_tracker components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from device_tracker component.

Swamp-Ig committed:
Moved fan components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from fan component.

Swamp-Ig committed:
Moved geo_location components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from geo_location component.

Swamp-Ig committed:
Moved image_processing components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from image_processing component.

Swamp-Ig committed:
Moved light components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from light component.

Swamp-Ig committed:
Moved lock components with tests into platform dirs.

Swamp-Ig committed:
Moved media_player components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from media_player component.

Swamp-Ig committed:
Moved scene components with tests into platform dirs.

Swamp-Ig committed:
Moved sensor components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from sensor component.

Swamp-Ig committed:
Moved switch components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from sensor component.

Swamp-Ig committed:
Moved vacuum components with tests into platform dirs.

Swamp-Ig committed:
Updated tests from vacuum component.

Swamp-Ig committed:
Moved weather components with tests into platform dirs.

OttoWinter commented:
Please remove all "For more details ..." from this PR. Some time ago it was decided that the links should be removed from source files (though it was probably only changed in src, not test files)

balloob commented:
My bad, I've updated my script in the other PR.

balloob commented:
From my experience, the first batch of errors can be found by running `script/gen_requirements_all.py`, as it imports all files.

Swamp-Ig committed:
Fixed __init__.py files

Swamp-Ig committed:
Fixes for stuff moved as part of this branch.

Swamp-Ig committed:
Fix stuff needed to merge with balloob's branch.

Swamp-Ig committed:
Formatting issues.

Swamp-Ig committed:
Missing __init__.py files.

Swamp-Ig committed:
Fix-ups

Swamp-Ig committed:
Fixup

Swamp-Ig committed:
Regenerated requirements.

Swamp-Ig committed:
Linting errors fixed.

Swamp-Ig committed:
Fixed more broken tests.

Swamp-Ig committed:
Missing init files.

Swamp-Ig committed:
Fix broken tests.

Swamp-Ig committed:
More broken tests

Swamp-Ig committed:
There seems to be a thread race condition.
I suspect the logger stuff is running in another thread, which means waiting until the aio loop is done is missing the log messages.
Used sleep instead because that allows the logger thread to run. I think the api_streams sensor might not be thread safe.

Swamp-Ig committed:
Disabled tests, will remove sensor in #22147

Swamp-Ig commented:
I disabled the tests and pushed just to get all the tests passing in tox.

Swamp-Ig commented:
Coverage is only decreased due to that test being disabled. The sensor is going to be removed anyhow.

Swamp-Ig commented:
I have a patch for coveragerc and codeowners. Just checking it before I commit.

balloob commented:
If you disable the test of api_streams, add it to coveragerc to exclude it from coverage.

Swamp-Ig committed:
Updated coverage and codeowners.

Swamp-Ig commented:
Commenting here to keep it somewhere...

The following script finds several files in `.coverageac` that don't exist anywhere the code base. 

## Script
```sh
for i in $(cat .coveragerc | tail +13 | head -681)
do 
    if [ $(ls $i 2>/dev/null | wc -l) -eq 0 ]
    then 
         echo $i
    fi
done

##Output post this patch:

```
homeassistant/components/apiai/*
homeassistant/components/nadtcp/media_player.py
homeassistant/components/notify/telstra.py
homeassistant/components/domain_expiry/sensor.py
homeassistant/components/netdata_public/sensor.py
```

## Output pre this patch:

```
homeassistant/components/apiai/*
homeassistant/components/camera/bloomsky.py
homeassistant/components/climate/homematic.py
homeassistant/components/climate/knx.py
homeassistant/components/cover/homematic.py
homeassistant/components/cover/knx.py
homeassistant/components/cover/rpi_gpio.py
homeassistant/components/cover/scsgate.py
homeassistant/components/device_tracker/tado.py
homeassistant/components/fan/wemo.py
homeassistant/components/light/hue.py
homeassistant/components/light/tplink.py
homeassistant/components/light/tradfri.py
homeassistant/components/media_player/nadtcp.py
homeassistant/components/notify/homematic.py
homeassistant/components/notify/telstra.py
homeassistant/components/remote/harmony.py
homeassistant/components/remote/itach.py
homeassistant/components/sensor/domain_expiry.py
homeassistant/components/sensor/netdata_public.py
homeassistant/components/switch/tplink.py
```

# Output at 0.89.1
```
homeassistant/components/apiai/*
homeassistant/components/camera/bloomsky.py
homeassistant/components/climate/homematic.py
homeassistant/components/climate/knx.py
homeassistant/components/cover/homematic.py
homeassistant/components/cover/knx.py
homeassistant/components/cover/rpi_gpio.py
homeassistant/components/cover/scsgate.py
homeassistant/components/device_tracker/tado.py
homeassistant/components/fan/wemo.py
homeassistant/components/light/hue.py
homeassistant/components/light/tplink.py
homeassistant/components/light/tradfri.py
homeassistant/components/media_player/nadtcp.py
homeassistant/components/notify/homematic.py
homeassistant/components/notify/telstra.py
homeassistant/components/opencv/*
homeassistant/components/raspyrfm/*
homeassistant/components/remote/harmony.py
homeassistant/components/remote/itach.py
homeassistant/components/sensor/domain_expiry.py
homeassistant/components/sensor/netdata_public.py
homeassistant/components/switch/tplink.py
```

Swamp-Ig commented:
OK I think this is ready to be reviewed now.

balloob reviewed (approved):
SHIP IT

Swamp-Ig commented:
Another comment:

```
$ ls $(cat ../izone_custom_component/movers.txt) | grep -v __init__.py | grep -v __pycache__ | grep -v services.yaml
homeassistant/components/air_quality:

homeassistant/components/alarm_control_panel:
manual_mqtt.py
manual.py

homeassistant/components/binary_sensor:

homeassistant/components/calendar:
demo.py

homeassistant/components/camera:
demo_0.jpg
demo_1.jpg
demo_2.jpg
demo_3.jpg
demo.py

homeassistant/components/climate:
const.py
demo.py
generic_thermostat.py
reproduce_state.py

homeassistant/components/cover:
demo.py
group.py

homeassistant/components/device_tracker:

homeassistant/components/fan:
demo.py

homeassistant/components/geo_location:
demo.py

homeassistant/components/image_processing:

homeassistant/components/light:
demo.py
group.py
light_profiles.csv
switch.py

homeassistant/components/lock:
demo.py

homeassistant/components/mailbox:

homeassistant/components/media_player:
const.py
demo.py
reproduce_state.py

homeassistant/components/remote:

homeassistant/components/scene:
homeassistant.py

homeassistant/components/sensor:
mold_indicator.py
strings.moon.json
strings.season.json

homeassistant/components/switch:

homeassistant/components/vacuum:
demo.py

homeassistant/components/water_heater:
demo.py

homeassistant/components/weather:
```
